Homework 4: Sentiment Analysis - Task 3
----

Names & Sections
----
Names: Dave Budhram(4120) and Akshay Dupuguntla(4120) (Write these in every notebook you submit. For each partner, write down whether you are a 4120 or a 6120 student.)

Task 3: Train a Logistic Regression Model (20 points)
----

Using `sklearn`'s implementation of `LogisticRegression`, conduct a similar analysis on the performance of a Logistic Regression classifier on the provided data set.

Using the `time` module, you'll compare and contrast how long it takes your home-grown BoW vectorizing function vs. `sklearn`'s `CountVectorizer`.


In [4]:
from sklearn.linear_model import LogisticRegression
# https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html
from sklearn.feature_extraction.text import CountVectorizer
from nltk.metrics.scores import (precision, recall, f_measure, accuracy)
from collections import Counter
import time
import sentiment_utils as sutils
from sentiment_utils import convert2DArray
from sentiment_utils import CustomCountVectorizer

In [5]:
# define constants for the files we are using
TRAIN_FILE = "movie_reviews_train.txt"
DEV_FILE = "movie_reviews_dev.txt"

# load in your data and make sure you understand the format
# Do not print out too much so as to impede readability of your notebook
X_train, y_train = sutils.generate_tuples_from_file(TRAIN_FILE)
X_dev, y_dev = sutils.generate_tuples_from_file(DEV_FILE)

# some variables you may want to use
BINARIZED = True
USE_COUNT_VECTORIZER = False


In [3]:
# Write the functions needed (here or in sentiment_utils.py) to create vectorized BoW representations
# of your data. We recommend starting with a multinomial BoW representation.
# Each training example should be represented as a sparse vector.

X_train_strs = convert2DArray(X_train)
X_dev_strs = convert2DArray(X_dev) 
count_vectorizer = CountVectorizer(binary=False)
count_vectorizer.fit(X_train_strs)
X_train_bow = count_vectorizer.transform(X_train_strs)
X_dev_bow = count_vectorizer.transform(X_dev_strs)

clf = LogisticRegression()
clf.fit(X_train_bow, y_train)
y_pred = clf.predict(X_dev_bow)

accuracy_val = accuracy(y_dev, y_pred)

# Calculate precision, recall, and F1 score
print("Accuracy:", accuracy_val)

Accuracy: 0.78


/opt/homebrew/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [123]:


customCountVectorizer = CustomCountVectorizer(binary=False)
customCountVectorizer.fit(X_train)
X_train_bow = customCountVectorizer.transform(X_train)
X_dev_bow = customCountVectorizer.transform(X_dev)
# print(matrix[0])
# print(len(customCountVectorizer.vocab))

clf = LogisticRegression()
clf.fit(X_train_bow, y_train)
y_pred = clf.predict(X_dev_bow)

accuracy_val = accuracy(y_dev, y_pred)



# all_reviews = []
# for review in train_tups_x:
#   all_reviews.extend(review)
# vocab = set(all_reviews)
# count = Counter(all_reviews)
# print(len(set(count.keys())))
# vocab = count.keys()
# print(len(count)) # number of words 
# for review in train_tups_x:

  

/opt/homebrew/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [6]:
# how much time does it take to featurize the all data with your implementation?

start = time.time()

# YOUR CODE HERE
custom_count_vectorizer = CustomCountVectorizer(binary=False)
custom_count_vectorizer.fit(X_train)
X_train_bow = custom_count_vectorizer.transform(X_train)
X_dev_bow = custom_count_vectorizer.transform(X_dev)

end = time.time()
print("That took:", end - start, "seconds")
print("Custom Count Vectorize Vocab Size:", len(custom_count_vectorizer.vocab))


That took: 0.3790550231933594 seconds
Custom Count Vectorize Vocab Size: 30705


In [8]:
# how much time does it take to featurize the all data with sklearn's CountVectorizer?

X_train_strs = convert2DArray(X_train)
X_dev_strs = convert2DArray(X_dev) 

start = time.time()

# YOUR CODE HERE
count_vectorizer = CountVectorizer(binary=False)
count_vectorizer.fit(X_train_strs)
X_train_bow = count_vectorizer.transform(X_train_strs)
# X_dev_bow = count_vectorizer.transform(X_dev_strs)
end = time.time()
print("That took:", end - start, "seconds")
print("Sklearn CountVectorizer Vocab Size:", len(count_vectorizer.vocabulary_))

That took: 0.3257288932800293 seconds
Sklearn CountVectorizer Vocab Size: 22596


1. How big is your vocabulary using your vectorization function(s)? 30705
2. How big is your vocabulary using the `CountVectorizer`? 22596

The difference in vocabulary is primarily attributed to the fact that our custom count vectorizer does not convert words to lowercase before adding them to the vocabulary. This means that words in different cases, such as "Word" and "word," are treated as separate terms and are included in the vocabulary as distinct features.

In [ ]:
#  write any code you need analyze the relative sparsity of your vectorized representations of the data

# YOUR CODE HERE

# Print out the average % of entries that are zeros in each vector in the vectorized training data
# YOUR CODE HERE

In [ ]:
# Using the provided dev set, evaluate your model with precision, recall, and f1 score as well as accuracy
# You may use nltk's implemented `precision`, `recall`, `f_measure`, and `accuracy` functions
# (make sure to look at the documentation for these functions!)
# you will be creating a similar graph for logistic regression and neural nets, so make sure
# you use functions wisely so that you do not have excessive repeated code
# write any helper functions you need in sentiment_utils.py (functions that you'll use in your other notebooks as well)


# create a graph of your classifier's performance on the dev set as a function of the amount of training data
# the x-axis should be the amount of training data (as a percentage of the total training data)
# the y-axis should be the performance of the classifier on the dev set
# the graph should have 4 lines, one for each of precision, recall, f1, and accuracy
# the graph should have a legend, title, and axis labels

# takes approx 30 sec on Felix's computer

Test the following 4 combinations to determine which has the best final f1 score for your Logistic Regression model:
- your vectorized features, multinomial: __enter your final f1 score here__
- CountVectorizer features, multinomial: __enter your final f1 score here__
- your vectorized features, binarized: __enter your final f1 score here__
- CountVectorizer features, binarized: __enter your final f1 score here__

Produce your graph(s) for the combination with the best final f1 score.




6120 REQUIRED
----

Find the top 100 most important features to your Logistic Regression classifier when using 100% of the training data. To access the weights of your model, you can access the `model.coef_` attribute. You'll want to use a `StandardScalar` preprocessor. This will help us deal with the fact that we expect counts of certain words to be higher (e.g. stop words).

To find the importance of a feature, calculate the absolute value of each weight in the model, then order your features according to the absolute values of these weights. The feature with the heighest absolute value weight has the most importance.

Use __your__ (not CountVectorizer) multinomial vectors for this analysis.

In [ ]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler

In [ ]:
# YOUR CODE HERE
# train a model on the scaled inputs
# This takes Felix's computer about 6.5 sec to run




In [ ]:
# print out the top 20 most informative features according to this model


In [ ]:
# re-evalaute your LR model with inputs that have been filtered to only use the top 500 most informative features


In [ ]:
# create the same graph as before, but with the filtered inputs
